File: 1202-smallest-string-with-swaps/1202-smallest-string-with-swaps.cpp
class Solution {
public:
    string smallestStringWithSwaps(string s, vector<vector<int>>& pairs) {
        return solve(s, pairs);
    }
private:
    int n;
    vector<bool> visited;
    vector<int> indices;
    string str = "";
    vector<vector<int>> adj;
    
    void dfs(string& s, int idx) {
        visited[idx] = true;
        
        indices.push_back(idx);
        str += s[idx];
        
        for(auto nbr: adj[idx]) {
            if(visited[nbr] == false)
                dfs(s, nbr);
        }
    }
    
    string solve(string s, vector<vector<int>>& pairs) {
        n = s.length();
        visited.resize(this -> n, false);
        adj.resize(this -> n);
        for(auto pair: pairs) {
            adj[pair[0]].push_back(pair[1]);
            adj[pair[1]].push_back(pair[0]);
        }
        
        for(int i = 0; i < n; ++i) {
            if(!visited[i]) {
                str = "";
                indices.clear();
                
                dfs(s, i);
                
                sort(str.begin(), str.end());
                sort(indices.begin(), indices.end());
                
                for(int j = 0; j < indices.size(); ++j) {
                    s[indices[j]] = str[j];
                }
            }
        }
        return s;
    }
};

File: 1202-smallest-string-with-swaps/1202-smallest-string-with-swaps.py
class Solution:
    def dfs(self, s, idx):
        self.visited[idx] = True
        
        self.indices.append(idx)
        self.string.append(s[idx])
        
        for nbr in self.adj[idx]:
            if self.visited[nbr] == False:
                self.dfs(s, nbr)
        
    def smallestStringWithSwaps(self, s: str, pairs: List[List[int]]) -> str:
        
        n = len(s)
        self.visited = [False] * n
        self.adj = [list() for _ in range(n)]
        
        self.indices = []
        self.string = []
        
        # Building adjacency list
        for pair in pairs:
            self.adj[pair[0]].append(pair[1])
            self.adj[pair[1]].append(pair[0])
            
        ans = ['-'] * n
        for i in range(n):
            if self.visited[i] == False:
                self.indices = []
                self.string = []
                
                # finding connected components
                self.dfs(s, i)
                
                # Sorting one connected component
                self.indices.sort()
                self.string.sort()
                
                # Arranging the found letters on their indices
                for j in range(len(self.indices)):
                    ans[self.indices[j]] = self.string[j]
                    
        return ''.join(ans)
                
            
        